extensions-core/caffeine-cache/src/main/java/io/druid/client/cache/CaffeineCache.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.client.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.primitives.Chars;
+import com.google.common.primitives.Ints;
+import com.metamx.common.logger.Logger;
+import com.metamx.emitter.service.ServiceEmitter;
+import com.metamx.emitter.service.ServiceMetricEvent;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.lz4.LZ4FastDecompressor;
+
+public class CaffeineCache implements io.druid.client.cache.Cache
+{
+  private static final Logger log = new Logger(CaffeineCache.class);
+  private static final int FIXED_COST = 8; // Minimum cost in "weight" per entry;
+  private static final LZ4Factory LZ4_FACTORY = LZ4Factory.fastestInstance();
+  private static final LZ4FastDecompressor LZ4_DECOMPRESSOR = LZ4_FACTORY.fastDecompressor();
+  private static final LZ4Compressor LZ4_COMPRESSOR = LZ4_FACTORY.fastCompressor();
+
+  private final Cache<NamedKey, byte[]> cache;
+  private final AtomicReference<CacheStats> priorStats = new AtomicReference<>(null);
+  private final CaffeineCacheConfig config;
+
+
+  public static CaffeineCache create(final CaffeineCacheConfig config)
+  {
+    return create(config, config.createExecutor());
+  }
+
+  // Used in testing
+  public static CaffeineCache create(final CaffeineCacheConfig config, @Nullable final Executor executor)
+  {
+    Caffeine<Object, Object> builder = Caffeine.newBuilder().recordStats();
+    if (config.getExpireAfter() >= 0) {
+      builder
+          .expireAfterAccess(config.getExpireAfter(), TimeUnit.MILLISECONDS);
```

Comment:
Curious as to your intent here. It would seem one would use this or size, but I'm not sure its useful when combining. A time-to-live helps protect against stale data, which might be combined with automatic refreshing to avoid exposing a latency penalty for hot entries. A time-to-idle is a poor man's hint of recency/frequency to the eviction policy. It was free to implement and a common feature, but less obvious when useful.